File: simulation/outputInterface.py
import numpy as np

def computePerformance(currentIndex, updateInterval, countersEvolution, 
                      labels, assignments, accuracies):
    '''
    计算网络的性能

    输入参数：
        1) currentIndex: 当前图像的索引，表示当前处理到了第几张图片
        2) updateInterval: 更新间隔，即每隔多少张图像计算一次性能
        3) countersEvolution: 二维 NumPy 数组，记录了过去 "updateInterval" 个周期内spikes计数器的历史
           - 每行对应一个训练步骤
           - 每列对应输出层中的一个元素（神经元）
        4) labels: NumPy 数组，包含所有图像的真实标签
        5) assignments: NumPy 数组，为每个输出神经元分配一个标签，用于分类映射
        6) accuracies: 字符串列表，记录历史的准确率

    输出：
        accuracies: 更新后的字符串列表，包含历史的准确率

    函数逻辑：
        该函数在指定的更新间隔检查网络性能，通过分析spikes计数器来推测分类结果，并与真实标签对比
    '''

    # 检查是否到达更新间隔的末尾
    # currentIndex % updateInterval == 0 表示当前索引是更新间隔的倍数
    # currentIndex > 0 避免在第一次循环就触发计算
    if currentIndex % updateInterval == 0 and currentIndex > 0:
        
        # 初始化最大spikes计数数组，长度为 updateInterval，初始值为 0
        # 用于跟踪每个时间步的最大spikes数
        maxCount = np.zeros(updateInterval)

        # 初始化分类结果数组，默认值设为 -1，表示未分类
        # dtype=np.int32 确保整数类型
        classification = -1 * np.ones(updateInterval, dtype=np.int32)

        # 获取当前更新间隔内的真实标签序列
        # 从 labels 中提取最近 updateInterval 个标签
        labelsSequence = labels[currentIndex - updateInterval : currentIndex]

        # 遍历所有可能的标签（0 到 9，对应 MNIST 的 10 个类别）
        for label in range(10):
            
            # 计算与当前标签关联的神经元的spikes总数
            # assignments == label 找到分配给当前标签的神经元
            # countersEvolution[:, assignments == label] 提取这些神经元的spikes数据
            # np.sum(..., axis=1) 按行求和，得到每个时间步的总spikes数
            spikesCount = np.sum(countersEvolution[:, assignments == label], axis=1)

            # 找到spikes计数大于当前最大值的时间步
            # whereMaxSpikes 是一个布尔数组，标记哪些时间步的spikes数超过了 maxCount
            whereMaxSpikes = spikesCount > maxCount

            # 将这些时间步的分类结果更新为当前标签
            classification[whereMaxSpikes] = label

            # 更新最大spikes计数，仅更新超过当前最大值的时间步
            maxCount[whereMaxSpikes] = spikesCount[whereMaxSpikes]

        # 打印计算出的分类结果，便于调试和观察
        print(classification)

        # 计算准确率并更新 accuracies 列表
        accuracies = updateAccuracy(classification, labelsSequence, accuracies)

    # 返回更新后的准确率历史
    return accuracies




def updateAccuracy(classification, labelsSequence, accuracies):
    '''
    计算准确率并将其添加到准确率历史列表中。

    输入参数：
        1) classification: NumPy 数组，记录了过去 "updateInterval" 个周期内网络的分类结果
        2) labelsSequence: NumPy 数组，记录了过去 "updateInterval" 个周期内的真实标签
        3) accuracies: 字符串列表，记录历史的准确率

    输出：
        accuracies: 更新后的字符串列表，包含历史的准确率

    函数逻辑：
        通过比较分类结果和真实标签，计算正确率并格式化为百分比字符串
    '''

    # 计算分类结果与真实标签相符的时间步数量
    # np.where(classification == labelsSequence) 返回相符位置的索引
    # [0].size 获取这些位置的数量
    correct = np.where(classification == labelsSequence)[0].size

    # 计算准确率百分比并格式化为两位小数字符串
    # correct/classification.size 计算正确比例
    # *100 转换为百分比
    # "{:.2f}".format() 保留两位小数
    accuracies += ["{:.2f}".format(correct/classification.size*100) + "%"]
    
    # 生成并打印准确率字符串，便于观察
    accuracyString = "\nAccuracy: " + str(accuracies) + "\n"
    print(accuracyString)

    # 返回更新后的准确率历史
    return accuracies

File: src/simulation/outputInterface.rs
use ndarray::{Array2, Array1, Axis};
use ndarray::s; // Import the s macro
use ndarray::ArrayBase; // Import ArrayBase
use ndarray::Data; // Import Data trait
use ndarray::Dim; // Import Dim
use std::fmt::Write;

/// 根据脉冲计数器计算网络性能并更新准确率。
///
/// # 输入参数
/// - `current_index`: 当前图像的索引 (`usize`)。
/// - `update_interval`: 计算性能的图像间隔数 (`usize`)。
/// - `counters_evolution`: 最近 `update_interval` 周期内脉冲计数器的二维数组。
///   形状: `(update_interval, n_neurons)`，类型: `&Array2<i32>`。
/// - `labels`: 所有图像标签的一维数组，类型: `&Array1<i32>`。
/// - `assignments`: 每个输出神经元的标签分配一维数组，类型: `&Array1<i32>`。
/// - `accuracies`: 包含准确率历史的字符串向量可变引用 (`&mut Vec<String>`)。
///
/// # 输出
/// 更新后的 `accuracies` 向量，附加了新的准确率。
///
/// # 描述
/// 该函数每隔 `update_interval` 个图像通过分析输出层的脉冲计数器来评估网络性能。
/// 它基于脉冲计数最高的神经元对图像进行分类，并更新准确率历史记录。
pub fn compute_performance<'a>(
    current_index: usize,
    update_interval: usize,
    counters_evolution: &'a Array2<i32>,
    labels: &'a Array1<i32>,
    assignments: &'a Array1<i32>,
    accuracies: &'a mut Vec<String>,
) -> &'a mut Vec<String> {
    // 检查是否到了计算性能的时间
    if current_index % update_interval == 0 && current_index > 0 {
        // 使用全零初始化 max_count
        let mut max_count = Array1::<f64>::zeros(update_interval);
        // 使用 -1 初始化 classification
        let mut classification = Array1::<i32>::from_elem(update_interval, -1);

        // 提取最近 update_interval 个图像的标签序列
        let labels_sequence = labels.slice(s![current_index - update_interval..current_index]);

        // 遍历每个可能的标签 (0 到 9)
        for label in 0..10 {
            // 创建一个掩码，where assignments 等于当前标签
            let mask = assignments.mapv(|a| a == label);
            // 选择具有此标签的神经元并沿轴 1 求和它们的脉冲计数
            let indices: Vec<usize> = mask
                .iter()
                .enumerate()
                .filter(|&(_, &b)| b)
                .map(|(i, _)| i)
                .collect();
            let spikes_count = counters_evolution
                .select(Axis(1), &indices)
                .sum_axis(Axis(1))
                .mapv(|x| x as f64);

            // 查找 spikes_count 超过 max_count 的位置 (element-wise comparison)
            let where_max_spikes = Array1::from_shape_fn(update_interval, |i| {
                spikes_count[i] > max_count[i]
            });

            // 在 spikes_count 较大时更新 classification 和 max_count
            for (i, &cond) in where_max_spikes.iter().enumerate() {
                if cond {
                    classification[i] = label;
                    max_count[i] = spikes_count[i];
                }
            }
        }

        // 打印计算出的分类结果
        println!("{:?}", classification);

        // 更新准确率历史
        update_accuracy(&classification, &labels_sequence, accuracies);
    }

    accuracies
}

/// 根据分类结果更新准确率历史。
///
/// # 输入参数
/// - `classification`: 最近 `update_interval` 周期内网络分类结果的一维数组，
///   类型: `&Array1<i32>`。
/// - `labels_sequence`: 最近 `update_interval` 周期内真实标签的一维数组，
///   类型: `&ArrayBase<impl Data<Elem=i32>, Dim<[usize; 1]>>`。
/// - `accuracies`: 包含准确率历史的字符串向量可变引用 (`&mut Vec<String>`)。
///
/// # 输出
/// 更新后的 `accuracies` 向量，附加了新的准确率。
///
/// # 描述
/// 该函数计算正确分类的百分比，并将其作为格式化字符串附加到准确率历史中。
fn update_accuracy<'a>(
    classification: &'a Array1<i32>,
    labels_sequence: &'a ArrayBase<impl Data<Elem=i32>, Dim<[usize; 1]>>, // Accept view type
    accuracies: &'a mut Vec<String>,
) -> &'a mut Vec<String> {
    // 统计正确分类的个数
    let correct = classification
        .iter()
        .zip(labels_sequence.iter())
        .filter(|(&c, &l)| c == l)
        .count();

    // 计算准确率百分比
    let accuracy = (correct as f64 / classification.len() as f64) * 100.0;

    // 将格式化的准确率附加到列表
    accuracies.push(format!("{:.2}%", accuracy));

    // 打印准确率历史
    let mut accuracy_string = String::new();
    write!(&mut accuracy_string, "\nAccuracy: {:?}\n", accuracies).unwrap();
    print!("{}", accuracy_string);

    accuracies
}

// 测试模块
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_compute_performance() {
        // 测试数据
        let update_interval = 5;
        let current_index = 5; // 恰好达到 update_interval
        let n_neurons = 10;

        // 模拟 counters_evolution: 5 个时间步，10 个神经元
        let counters_evolution = Array2::from_shape_vec(
            (update_interval, n_neurons),
            vec![
                1, 0, 0, 0, 0, 0, 0, 0, 0, 0, // 时间步 0
                0, 2, 0, 0, 0, 0, 0, 0, 0, 0, // 时间步 1
                0, 0, 3, 0, 0, 0, 0, 0, 0, 0, // 时间步 2
                0, 0, 0, 4, 0, 0, 0, 0, 0, 0, // 时间步 3
                0, 0, 0, 0, 5, 0, 0, 0, 0, 0, // 时间步 4
            ],
        )
        .unwrap();

        // 模拟 labels: 前 5 个标签
        let labels = Array1::from_vec(vec![0, 1, 2, 3, 4]);

        // 模拟 assignments: 每个神经元分配给对应的标签 (0 到 9)
        let assignments = Array1::from_vec((0..10).collect());

        // 初始化 accuracies
        let mut accuracies = Vec::new();

        // 调用 compute_performance (ignore unused result)
        let _result = compute_performance(current_index, update_interval, &counters_evolution, &labels, &assignments, &mut accuracies);

        // 验证结果 (use accuracies directly after mutable borrow ends)
        assert_eq!(accuracies.len(), 1); // 应该添加一个准确率
        assert!(accuracies[0].ends_with("%")); // 确保格式正确
        assert_eq!(accuracies[0], "100.00%"); // 由于每个时间步的脉冲计数最大值对应正确标签，准确率应为 100%

        // 验证 classification (预期应为 [0, 1, 2, 3, 4])
        let mut classification = Array1::<i32>::from_elem(update_interval, -1);
        let mut max_count = Array1::<f64>::zeros(update_interval);
        for label in 0..10 {
            let mask = assignments.mapv(|a| a == label);
            let indices: Vec<usize> = mask.iter().enumerate().filter(|&(_, &b)| b).map(|(i, _)| i).collect();
            let spikes_count = counters_evolution.select(Axis(1), &indices).sum_axis(Axis(1)).mapv(|x| x as f64);
            let where_max_spikes = Array1::from_shape_fn(update_interval, |i| {
                spikes_count[i] > max_count[i]
            });
            for (i, &cond) in where_max_spikes.iter().enumerate() {
                if cond {
                    classification[i] = label;
                    max_count[i] = spikes_count[i];
                }
            }
        }
        let expected_classification = Array1::from_vec(vec![0, 1, 2, 3, 4]);
        assert_eq!(classification, expected_classification);
    }

    #[test]
    fn test_update_accuracy() {
        // 测试数据
        let classification = Array1::from_vec(vec![0, 1, 2, 3, 4]); // 正确分类
        let labels_sequence = Array1::from_vec(vec![0, 1, 2, 3, 4]); // 真实标签
        let mut accuracies = Vec::new();

        // 调用 update_accuracy
        let _result = update_accuracy(&classification, &labels_sequence, &mut accuracies);

        // 验证结果
        assert_eq!(accuracies.len(), 1); // 应该添加一个准确率
        assert_eq!(accuracies[0], "100.00%"); // 所有分类正确，准确率应为 100%

        // 测试部分正确的情况
        let classification_partial = Array1::from_vec(vec![0, 1, 2, 3, 5]); // 最后一个错误
        let labels_sequence_partial = Array1::from_vec(vec![0, 1, 2, 3, 4]);
        let mut accuracies_partial = Vec::new();
        let _result_partial = update_accuracy(&classification_partial, &labels_sequence_partial, &mut accuracies_partial);

        assert_eq!(accuracies_partial.len(), 1);
        assert_eq!(accuracies_partial[0], "80.00%"); // 4/5 正确，准确率应为 80%
    }
}